### DISCUSSION TITLE
Mach3 usb with 2nd probe [resolved]

### DISCUSSION THREAD

**Discussion by rmaltais** [grblHAL/core#496] [Q&A] [answered]
Hi,

Why i can't build a firmware with 2nd probe? It is possible to use a door input for this probe because i don't use it. Can i change firmware for use this plug-in?

![msedge_Sb7w9TvTHd](https://github.com/grblHAL/core/assets/41405404/20d7fd99-9591-4344-984c-5d45b91ecec6)

thank

**Comment by nickshl** [comment]
Why do you need it? Normally Open probes can be easily connected in parallel. Normally Closed more difficult - electronic probes can't be connected in series. However if one probe is electronic(_like 3D touch probe_) and other probe is just a switch(_like tool setter_) those probes can be connected in series(_output of 3D touch probe trough switch in toolsetter to controller input_). But in this case both probes have to be connected to do measurements. If you don't need to have sensors connected whole time, it easier to use GX12 connector and connect particular probe before measurement.
The biggest problem with more that one probe input - GCode doesn't support that. As result there no way to use two inputs with NC sensors when one sensor is disconnected, but when I want to make measurement with another. There is no way to tell grblHAL "_Ignore first probe, and use only the second_".

**Reply by rmaltais** [reply]
Hi,

The problem is that my tool probe is NC and my electronic probe is a NO signal. I have to reverse the signal in the controller every time I want to change the probe. I think I will have to install a relay on my tool setter to invert the signal and connect it in parallel. If not, is there another solution?

Regards,

**Reply by nickshl** [reply, marked as answer]
Of course there is another solution. You need two parts: NPN transistor(2N2222, BC547 or similar) and resistor.
![IMG_20240424_111926687~2](https://github.com/grblHAL/core/assets/38679154/d33466c2-bc01-46cc-9b58-8ac1db7b4774)
This schematic converts NO sensor to NC sensor. Current trough base should be about 5-10-15 mA. You can use GX16 connectors(_relatively cheap on AliExpress_) - those big enough and you should be able to place transistor with TO-92 package and resistor inside connector(_on a sensor side_).

**Reply by terjeio** [reply]
I am working on a solution to this, by defining optional inputs as aux inputs and claiming them only when configured. The Web Builder has to be updated to keep track of aux pin usage - on my todo list.

**Reply by rmaltais** [reply]
Thank for all your work, very appreciate.

**Reply by rmaltais** [reply]
Hi,

I finally made the assembly with a BC547 and a 22K resistor. It doesn't work and I tried a 47K resistor without success. In fact, if I test the assembly outside of my circuit, it works on my multimeter and I have a 100% open circuit. If I place the same circuit inside with the PSU which also powers my probe it makes a weak link which closes does not fully open the circuit so it always remains active. A very weak LED tells me that the state is active.

**Reply by nickshl** [reply]
What is your sensor supply voltage? The resistor value seems way high to me. For 5 mA base current at 5V you need resistor  5 / 0.005 = 1k, at 12V 12 / 0.005 = 2.4k.

**Reply by rmaltais** [reply]
> What is your sensor supply voltage? The resistor value seems way high to me. For 5 mA base current at 5V you need resistor  5 / 0.005 = 1k, at 12V 12 / 0.005 = 2.4k.

My supply is 24V.

**Reply by nickshl** [reply]
Try 2.4k or so - base current should be 10 mA in this case.

**Reply by rmaltais** [reply]
Hi again,
It works now but I found the problem. It's not really the resistor value but rather my circuit in general which all shares the same PSU. If I just unplug the cable for my 3D probe, it works. If I plug in my 3D probe, it also works. If I leave the 3D probe cable but without the probe at the other end, it doesn't work and the probe triggers for no immediate reason. I investigate about that.

**Reply by rmaltais** [reply]
I finally found the solution to my installation-related instability problem. By adding a simple 1N914 diode to the emitter, everything works perfectly and there are no more reading errors. Also, even if the 3D probe is not connected, I no longer have to disconnect my cable.
![Schematic_Probe-inversion](https://github.com/grblHAL/core/assets/41405404/e95c1367-e5d2-4cb8-9fff-900b5d3ca6db)

**Comment by nickshl** [comment]
> I am working on a solution to this, by defining optional inputs as aux inputs and claiming them only when configured.

Is there any plans to add which sensor to use? Something like G38.2.N, where N - is probe number, and if it omitted(_for backward compatibility_), then all probes is monitoring. Or any other way? I don't want to have NC 3D touch probe always connected, but I have to in order to use tool length sensor...

**Reply by terjeio** [reply]
> Something like G38.2.N, where N - is probe number, and if it omitted(for backward compatibility), 

This is non-standard and will require changes to the parser since 38.2.N cannot be converted to a float - so I say no.

> Or any other way?

You have shown one option above, [plugin code that adds a M-code](https://github.com/grblHAL/Templates/blob/master/my_plugin/Probe_select/my_plugin.c) or switches probe input automatically is another. Automatic switching is possible if a toolsetter is located at a fixed position (G59.3 is used to specify that) - an [event is generated](https://github.com/grblHAL/core/blob/3a7a65ea7c3312987cd7b4e7b9552e63e4639fb9/motion_control.c#L995-L998) if probing is within range.
